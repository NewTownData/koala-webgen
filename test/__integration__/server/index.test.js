const fs = require('fs');
const path = require('path');
const http = require('http');
const { TextDecoder } = require('util');
const server = require('../../../src/server');
const init = require('../../../src/init');
const now = require('../../../src/context/now');

jest.mock('../../../src/context/now');

const tempRoot = path.join(__dirname, '.temp_server');
const websitePath = path.join(tempRoot, 'website');

const port = 3082;
const urlPrefix = `http://localhost:${port}`;

const TestDate = new Date(Date.UTC(2020, 11, 1, 14, 10, 20));

function utf8(buffer) {
  const decoder = new TextDecoder('utf-8');
  return decoder.decode(buffer);
}

function fetch(url) {
  return new Promise((resolve, reject) => {
    const req = http.get(url);
    req.on('response', (res) => {
      const chunks = [];
      res.on('data', (chunk) => {
        chunks.push(chunk);
      });
      res.on('end', async () => {
        try {
          const data = Buffer.concat(chunks);
          resolve({
            statusCode: res.statusCode,
            headers: res.headers,
            data,
          });
        } catch (e) {
          reject(e);
        }
      });
    });
    req.on('error', (err) => {
      reject(err);
    });
    req.on('timeout', () => {
      req.abort();
    });
  });
}

function modifyConfiguration(configPath) {
  const config = JSON.parse(fs.readFileSync(configPath, { encoding: 'utf-8' }));
  const updatedConfig = {
    ...config,
    websitePath: '/prefix/',
    menu: [{ name: 'About', link: '/prefix/about.html' }],
  };
  fs.writeFileSync(configPath, JSON.stringify(updatedConfig), {
    encoding: 'utf-8',
  });
}

describe('server', () => {
  let srv = null;

  beforeAll(async () => {
    fs.mkdirSync(tempRoot, { recursive: true });

    init(websitePath);
    modifyConfiguration(path.join(websitePath, 'configuration.json'));
    srv = await server(websitePath, port);

    now.mockClear();
    now.mockImplementation(() => TestDate);
  });

  afterAll((done) => {
    fs.rmSync(tempRoot, { recursive: true });
    srv.close(() => {
      done();
    });
  });

  it('redirect / to /prefix/', async () => {
    const response = await fetch(`${urlPrefix}/`);
    expect(response.statusCode).toBe(302);
    expect(response.headers.location).toBe('/prefix/');
  });

  it('index', async () => {
    const response = await fetch(`${urlPrefix}/prefix/`);
    expect(response.statusCode).toBe(200);
    expect(response.headers['content-type']).toBe('text/html; charset=utf-8');

    expect(utf8(response.data)).toMatchSnapshot();
  });

  it('404', async () => {
    const response = await fetch(`${urlPrefix}/prefix/xyz.txt`);
    expect(response.statusCode).toBe(404);
  });

  it('page about.html', async () => {
    const response = await fetch(`${urlPrefix}/prefix/about.html`);
    expect(response.statusCode).toBe(200);
    expect(response.headers['content-type']).toBe('text/html; charset=utf-8');
    expect(utf8(response.data)).toMatchSnapshot();
  });

  it('post test/test.html', async () => {
    const response = await fetch(`${urlPrefix}/prefix/test/test.html`);
    expect(response.statusCode).toBe(200);
    expect(response.headers['content-type']).toBe('text/html; charset=utf-8');
    expect(utf8(response.data)).toMatchSnapshot();
  });

  it('tags tags/hello/page-1.html', async () => {
    const response = await fetch(`${urlPrefix}/prefix/tags/hello/page-1.html`);
    expect(response.statusCode).toBe(200);
    expect(response.headers['content-type']).toBe('text/html; charset=utf-8');
    expect(utf8(response.data)).toMatchSnapshot();
  });

  it('posts page-2.html', async () => {
    const response = await fetch(`${urlPrefix}/prefix/posts/page-2.html`);
    expect(response.statusCode).toBe(200);
    expect(response.headers['content-type']).toBe('text/html; charset=utf-8');
    expect(utf8(response.data)).toMatchSnapshot();
  });

  it('archive /archive/2019-03/page-1.html', async () => {
    const response = await fetch(
      `${urlPrefix}/prefix/archive/2019-03/page-1.html`,
    );
    expect(response.statusCode).toBe(200);
    expect(response.headers['content-type']).toBe('text/html; charset=utf-8');
    expect(utf8(response.data)).toMatchSnapshot();
  });

  it('500', async () => {
    const response = await fetch(`${urlPrefix}/prefix/%%%`);
    expect(response.statusCode).toBe(500);

    expect(utf8(response.data)).toEqual(
      'Internal Server Error\nError: URI malformed',
    );
  });

  it('feed.rss.xml', async () => {
    const response = await fetch(`${urlPrefix}/prefix/feed.rss.xml`);
    expect(response.statusCode).toBe(200);
    expect(response.headers['content-type']).toBe('text/xml');
    expect(utf8(response.data)).toMatchSnapshot();
  });
});
