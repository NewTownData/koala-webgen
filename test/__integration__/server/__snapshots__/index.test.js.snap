// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server archive /archive/2019-03/page-1.html 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\" />
  <title>Archive March 2019 - Koala</title>
  <meta name=\\"robots\\" content=\\"index, follow\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, user-scalable=yes\\" />
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\" />
  <link rel=\\"canonical\\" href=\\"https://www.example.com/prefix/archive/2019-03/page-1.html\\" />
  <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/prefix/style.css\\" />
  <link rel=\\"apple-touch-icon\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"icon\\" sizes=\\"512x512\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"shortcut icon\\" href=\\"/prefix/favicon.ico\\" />
  <meta name=\\"description\\" content=\\"A simple static website generator\\" />
</head>

<body>
  <header>
  <div class=\\"header-logo-panel\\">
    <div class=\\"header-logo\\">
      <div class=\\"header-image\\">
        <a href=\\"/prefix/\\">
          <img src=\\"/prefix/images/logo-128.png\\" alt=\\"Koala Logo\\" />
        </a>
      </div>
      <div class=\\"header-text\\">
        <div class=\\"header-title\\">
          <a href=\\"/prefix/\\">Koala</a>
        </div>
        <div class=\\"header-subtitle\\">
          <a href=\\"/prefix/\\">Static Website Generator</a>
        </div>
      </div>
    </div>
  </div>

  <nav>
    <ul>
      <li><a href=\\"/prefix/about.html\\">About</a></li>
    </ul>
  </nav>
</header>
  <main>
  <div class=\\"pages\\">
    <h1>Archive March 2019</h1>
    <section>
      <h2>
        <a href=\\"/prefix/post5.html\\">Post 5</a>
      </h2>
      <div class=\\"post-date\\">3 March 2019</div>
      <p>
  Content 5
</p>
      <a href=\\"/prefix/post5.html\\">Read More</a>
    </section>
    <section>
      <h2>
        <a href=\\"/prefix/post4.html\\">Post 4</a>
      </h2>
      <div class=\\"post-date\\">2 March 2019</div>
      <p>
  Content 4
</p>
      <a href=\\"/prefix/post4.html\\">Read More</a>
    </section>
    <section>
      <h2>
        <a href=\\"/prefix/post3.html\\">Post 3</a>
      </h2>
      <div class=\\"post-date\\">1 March 2019</div>
      <p>
  Content 3
</p>
      <a href=\\"/prefix/post3.html\\">Read More</a>
    </section>
  </div>
  <nav class=\\"pages-navigation\\">
    <div class=\\"page-previous\\">
    </div>
    <div class=\\"page-number\\">Page 1</div>
    <div class=\\"page-next\\">
    </div>
  </nav>



</main>
  <div class=\\"navigation\\">
</div>

<aside>
  <h2>Archives</h2>
  <ul>
    <li>
      <a href=\\"/prefix/archive/2020-06/page-1.html\\">June 2020</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-03/page-1.html\\">March 2019</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-01/page-1.html\\">January 2019</a>
    </li>
  </ul>
</aside>

<aside>
  <h2>Tags</h2>
  <ul>
    <li>
      <a href=\\"/prefix/tags/hello/page-1.html\\">hello</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/intro/page-1.html\\">intro</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/number/page-1.html\\">number</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/test/page-1.html\\">test</a>
    </li>
  </ul>
</aside>

  <footer>
  <b>Koala</b><br />
  Static Website Generator<br />
  <a href=\\"https://www.example.com/prefix/\\">https://www.example.com/prefix/</a>
</footer>
</body>

</html>"
`;

exports[`server feed.rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" ?>
<rss version=\\"2.0\\">
  <channel>
    <title>Koala - Static Website Generator</title>
    <description>A simple static website generator</description>
    <link>https://www.example.com/prefix/</link>
    <lastBuildDate>Tue, 01 Dec 2020 14:10:20 GMT</lastBuildDate>
    <pubDate>Tue, 01 Dec 2020 14:10:20 GMT</pubDate>
    <item>
      <title>Hello World</title>
      <description>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet, diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan. Aenean eu lacus quam. Morbi consequat ante quis eros iaculis tempus. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Nulla sit amet urna vitae elit iaculis blandit. Mauris dignissim consectetur augue vel varius.</description>
      <link>https://www.example.com/prefix/hello-world.html</link>
      <guid>https://www.example.com/prefix/hello-world.html</guid>
      <pubDate>Sat, 20 Jun 2020 10:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Test</title>
      <description>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet, diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan.</description>
      <link>https://www.example.com/prefix/test/test.html</link>
      <guid>https://www.example.com/prefix/test/test.html</guid>
      <pubDate>Wed, 10 Jun 2020 10:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 5</title>
      <description>Content 5</description>
      <link>https://www.example.com/prefix/post5.html</link>
      <guid>https://www.example.com/prefix/post5.html</guid>
      <pubDate>Sun, 03 Mar 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 4</title>
      <description>Content 4</description>
      <link>https://www.example.com/prefix/post4.html</link>
      <guid>https://www.example.com/prefix/post4.html</guid>
      <pubDate>Sat, 02 Mar 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 3</title>
      <description>Content 3</description>
      <link>https://www.example.com/prefix/post3.html</link>
      <guid>https://www.example.com/prefix/post3.html</guid>
      <pubDate>Fri, 01 Mar 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 2</title>
      <description>Content 2</description>
      <link>https://www.example.com/prefix/post2.html</link>
      <guid>https://www.example.com/prefix/post2.html</guid>
      <pubDate>Wed, 02 Jan 2019 16:45:00 GMT</pubDate>
    </item>
    <item>
      <title>Post 1</title>
      <description>Content 1</description>
      <link>https://www.example.com/prefix/post1.html</link>
      <guid>https://www.example.com/prefix/post1.html</guid>
      <pubDate>Tue, 01 Jan 2019 16:45:00 GMT</pubDate>
    </item>
  </channel>
</rss>"
`;

exports[`server index 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\" />
  <title>Koala - Static Website Generator</title>
  <meta name=\\"robots\\" content=\\"index, follow\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, user-scalable=yes\\" />
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\" />
  <link rel=\\"canonical\\" href=\\"https://www.example.com/prefix/\\" />
  <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/prefix/style.css\\" />
  <link rel=\\"apple-touch-icon\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"icon\\" sizes=\\"512x512\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"shortcut icon\\" href=\\"/prefix/favicon.ico\\" />
  <meta name=\\"description\\" content=\\"A simple static website generator\\" />
</head>

<body>
  <header>
  <div class=\\"header-logo-panel\\">
    <div class=\\"header-logo\\">
      <div class=\\"header-image\\">
        <a href=\\"/prefix/\\">
          <img src=\\"/prefix/images/logo-128.png\\" alt=\\"Koala Logo\\" />
        </a>
      </div>
      <div class=\\"header-text\\">
        <div class=\\"header-title\\">
          <a href=\\"/prefix/\\">Koala</a>
        </div>
        <div class=\\"header-subtitle\\">
          <a href=\\"/prefix/\\">Static Website Generator</a>
        </div>
      </div>
    </div>
  </div>

  <nav>
    <ul>
      <li><a href=\\"/prefix/about.html\\">About</a></li>
    </ul>
  </nav>
</header>
  <main>
  <div class=\\"pages\\">
    <h1>Posts</h1>
    <section>
      <h2>
        <a href=\\"/prefix/hello-world.html\\">Hello World</a>
      </h2>
      <div class=\\"post-date\\">20 June 2020</div>
      <p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus
  dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet,
  diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan. Aenean eu
  lacus quam. Morbi consequat ante quis eros iaculis tempus. Orci varius natoque penatibus et magnis dis parturient
  montes, nascetur ridiculus mus. Nulla sit amet urna vitae elit iaculis blandit. Mauris dignissim consectetur augue vel
  varius.
</p>

      <a href=\\"/prefix/hello-world.html\\">Read More</a>
    </section>
    <section>
      <h2>
        <a href=\\"/prefix/test/test.html\\">Test</a>
      </h2>
      <div class=\\"post-date\\">10 June 2020</div>
      <p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus
  dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet,
  diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan.
</p>

      <a href=\\"/prefix/test/test.html\\">Read More</a>
    </section>
    <section>
      <h2>
        <a href=\\"/prefix/post5.html\\">Post 5</a>
      </h2>
      <div class=\\"post-date\\">3 March 2019</div>
      <p>
  Content 5
</p>
      <a href=\\"/prefix/post5.html\\">Read More</a>
    </section>
    <section>
      <h2>
        <a href=\\"/prefix/post4.html\\">Post 4</a>
      </h2>
      <div class=\\"post-date\\">2 March 2019</div>
      <p>
  Content 4
</p>
      <a href=\\"/prefix/post4.html\\">Read More</a>
    </section>
    <section>
      <h2>
        <a href=\\"/prefix/post3.html\\">Post 3</a>
      </h2>
      <div class=\\"post-date\\">1 March 2019</div>
      <p>
  Content 3
</p>
      <a href=\\"/prefix/post3.html\\">Read More</a>
    </section>
  </div>
  <nav class=\\"pages-navigation\\">
    <div class=\\"page-previous\\">
    </div>
    <div class=\\"page-number\\">Page 1</div>
    <div class=\\"page-next\\">
      <a href=\\"/prefix/posts/page-2.html\\">Next Page &#8594;</a>
    </div>
  </nav>



</main>
  <div class=\\"navigation\\">
</div>

<aside>
  <h2>Archives</h2>
  <ul>
    <li>
      <a href=\\"/prefix/archive/2020-06/page-1.html\\">June 2020</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-03/page-1.html\\">March 2019</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-01/page-1.html\\">January 2019</a>
    </li>
  </ul>
</aside>

<aside>
  <h2>Tags</h2>
  <ul>
    <li>
      <a href=\\"/prefix/tags/hello/page-1.html\\">hello</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/intro/page-1.html\\">intro</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/number/page-1.html\\">number</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/test/page-1.html\\">test</a>
    </li>
  </ul>
</aside>

  <footer>
  <b>Koala</b><br />
  Static Website Generator<br />
  <a href=\\"https://www.example.com/prefix/\\">https://www.example.com/prefix/</a>
</footer>
</body>

</html>"
`;

exports[`server page about.html 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\" />
  <title>About - Koala</title>
  <meta name=\\"robots\\" content=\\"index, follow\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, user-scalable=yes\\" />
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\" />
  <link rel=\\"canonical\\" href=\\"https://www.example.com/prefix/about.html\\" />
  <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/prefix/style.css\\" />
  <link rel=\\"apple-touch-icon\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"icon\\" sizes=\\"512x512\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"shortcut icon\\" href=\\"/prefix/favicon.ico\\" />
  <meta name=\\"description\\" content=\\"A simple static website generator\\" />
</head>

<body>
  <header>
  <div class=\\"header-logo-panel\\">
    <div class=\\"header-logo\\">
      <div class=\\"header-image\\">
        <a href=\\"/prefix/\\">
          <img src=\\"/prefix/images/logo-128.png\\" alt=\\"Koala Logo\\" />
        </a>
      </div>
      <div class=\\"header-text\\">
        <div class=\\"header-title\\">
          <a href=\\"/prefix/\\">Koala</a>
        </div>
        <div class=\\"header-subtitle\\">
          <a href=\\"/prefix/\\">Static Website Generator</a>
        </div>
      </div>
    </div>
  </div>

  <nav>
    <ul>
      <li><a href=\\"/prefix/about.html\\">About</a></li>
    </ul>
  </nav>
</header>
  <main>

  <section>
    <h1>
      <a href=\\"/prefix/about.html\\">About</a>
    </h1>
    <p>
  The project aims to create simple to use, scalable and fast static website generator.
</p>
<p>
  More info at <a href=\\"https://github.com/KoalaSpace/koala-webgen\\">GitHub</a>
</p>
  </section>


</main>
  <div class=\\"navigation\\">
</div>

<aside>
  <h2>Archives</h2>
  <ul>
    <li>
      <a href=\\"/prefix/archive/2020-06/page-1.html\\">June 2020</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-03/page-1.html\\">March 2019</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-01/page-1.html\\">January 2019</a>
    </li>
  </ul>
</aside>

<aside>
  <h2>Tags</h2>
  <ul>
    <li>
      <a href=\\"/prefix/tags/hello/page-1.html\\">hello</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/intro/page-1.html\\">intro</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/number/page-1.html\\">number</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/test/page-1.html\\">test</a>
    </li>
  </ul>
</aside>

  <footer>
  <b>Koala</b><br />
  Static Website Generator<br />
  <a href=\\"https://www.example.com/prefix/\\">https://www.example.com/prefix/</a>
</footer>
</body>

</html>"
`;

exports[`server post test/test.html 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\" />
  <title>Test - Koala</title>
  <meta name=\\"robots\\" content=\\"index, follow\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, user-scalable=yes\\" />
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\" />
  <link rel=\\"canonical\\" href=\\"https://www.example.com/prefix/test/test.html\\" />
  <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/prefix/style.css\\" />
  <link rel=\\"apple-touch-icon\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"icon\\" sizes=\\"512x512\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"shortcut icon\\" href=\\"/prefix/favicon.ico\\" />
  <meta name=\\"description\\" content=\\"A simple static website generator\\" />
</head>

<body>
  <header>
  <div class=\\"header-logo-panel\\">
    <div class=\\"header-logo\\">
      <div class=\\"header-image\\">
        <a href=\\"/prefix/\\">
          <img src=\\"/prefix/images/logo-128.png\\" alt=\\"Koala Logo\\" />
        </a>
      </div>
      <div class=\\"header-text\\">
        <div class=\\"header-title\\">
          <a href=\\"/prefix/\\">Koala</a>
        </div>
        <div class=\\"header-subtitle\\">
          <a href=\\"/prefix/\\">Static Website Generator</a>
        </div>
      </div>
    </div>
  </div>

  <nav>
    <ul>
      <li><a href=\\"/prefix/about.html\\">About</a></li>
    </ul>
  </nav>
</header>
  <main>


  <section>
    <h1>
      <a href=\\"/prefix/test/test.html\\">Test</a>
    </h1>
    <div class=\\"post-date\\">10 June 2020</div>
    <div class=\\"post-tags\\">
      <div class=\\"post-tags-name\\">Tags: </div>
      <ul>
        <li>
          <a href=\\"/prefix/tags/test/page-1.html\\">test</a>
        </li>
      </ul>
    </div>
    <p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus
  dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet,
  diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan.
</p>
<p>
  Aenean facilisis sit amet neque a ultricies. Sed consectetur lorem quis lobortis tincidunt. Interdum et malesuada
  fames ac ante ipsum primis in faucibus. Nullam at elit pulvinar, porta lorem sed, facilisis orci. Proin lacinia nisl
  at placerat ornare. Donec ultrices dui ac varius luctus. Mauris sagittis ornare arcu, a dictum ante aliquet id.
</p>
  </section>

</main>
  <div class=\\"navigation\\">
</div>

<aside>
  <h2>Archives</h2>
  <ul>
    <li>
      <a href=\\"/prefix/archive/2020-06/page-1.html\\">June 2020</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-03/page-1.html\\">March 2019</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-01/page-1.html\\">January 2019</a>
    </li>
  </ul>
</aside>

<aside>
  <h2>Tags</h2>
  <ul>
    <li>
      <a href=\\"/prefix/tags/hello/page-1.html\\">hello</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/intro/page-1.html\\">intro</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/number/page-1.html\\">number</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/test/page-1.html\\">test</a>
    </li>
  </ul>
</aside>

  <footer>
  <b>Koala</b><br />
  Static Website Generator<br />
  <a href=\\"https://www.example.com/prefix/\\">https://www.example.com/prefix/</a>
</footer>
</body>

</html>"
`;

exports[`server posts page-2.html 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\" />
  <title>Posts - Page 2 - Koala</title>
  <meta name=\\"robots\\" content=\\"index, follow\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, user-scalable=yes\\" />
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\" />
  <link rel=\\"canonical\\" href=\\"https://www.example.com/prefix/posts/page-2.html\\" />
  <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/prefix/style.css\\" />
  <link rel=\\"apple-touch-icon\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"icon\\" sizes=\\"512x512\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"shortcut icon\\" href=\\"/prefix/favicon.ico\\" />
  <meta name=\\"description\\" content=\\"A simple static website generator\\" />
</head>

<body>
  <header>
  <div class=\\"header-logo-panel\\">
    <div class=\\"header-logo\\">
      <div class=\\"header-image\\">
        <a href=\\"/prefix/\\">
          <img src=\\"/prefix/images/logo-128.png\\" alt=\\"Koala Logo\\" />
        </a>
      </div>
      <div class=\\"header-text\\">
        <div class=\\"header-title\\">
          <a href=\\"/prefix/\\">Koala</a>
        </div>
        <div class=\\"header-subtitle\\">
          <a href=\\"/prefix/\\">Static Website Generator</a>
        </div>
      </div>
    </div>
  </div>

  <nav>
    <ul>
      <li><a href=\\"/prefix/about.html\\">About</a></li>
    </ul>
  </nav>
</header>
  <main>
  <div class=\\"pages\\">
    <h1>Posts</h1>
    <section>
      <h2>
        <a href=\\"/prefix/post2.html\\">Post 2</a>
      </h2>
      <div class=\\"post-date\\">2 January 2019</div>
      <p>
  Content 2
</p>
      <a href=\\"/prefix/post2.html\\">Read More</a>
    </section>
    <section>
      <h2>
        <a href=\\"/prefix/post1.html\\">Post 1</a>
      </h2>
      <div class=\\"post-date\\">1 January 2019</div>
      <p>
  Content 1
</p>
      <a href=\\"/prefix/post1.html\\">Read More</a>
    </section>
  </div>
  <nav class=\\"pages-navigation\\">
    <div class=\\"page-previous\\">
      <a href=\\"/prefix/\\">&#8592; Previous Page</a>
    </div>
    <div class=\\"page-number\\">Page 2</div>
    <div class=\\"page-next\\">
    </div>
  </nav>



</main>
  <div class=\\"navigation\\">
</div>

<aside>
  <h2>Archives</h2>
  <ul>
    <li>
      <a href=\\"/prefix/archive/2020-06/page-1.html\\">June 2020</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-03/page-1.html\\">March 2019</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-01/page-1.html\\">January 2019</a>
    </li>
  </ul>
</aside>

<aside>
  <h2>Tags</h2>
  <ul>
    <li>
      <a href=\\"/prefix/tags/hello/page-1.html\\">hello</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/intro/page-1.html\\">intro</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/number/page-1.html\\">number</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/test/page-1.html\\">test</a>
    </li>
  </ul>
</aside>

  <footer>
  <b>Koala</b><br />
  Static Website Generator<br />
  <a href=\\"https://www.example.com/prefix/\\">https://www.example.com/prefix/</a>
</footer>
</body>

</html>"
`;

exports[`server tags tags/hello/page-1.html 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\" />
  <title>Tag hello - Koala</title>
  <meta name=\\"robots\\" content=\\"index, follow\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, user-scalable=yes\\" />
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\" />
  <link rel=\\"canonical\\" href=\\"https://www.example.com/prefix/tags/hello/page-1.html\\" />
  <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/prefix/style.css\\" />
  <link rel=\\"apple-touch-icon\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"icon\\" sizes=\\"512x512\\" href=\\"/prefix/images/logo-512.png\\" />
  <link rel=\\"shortcut icon\\" href=\\"/prefix/favicon.ico\\" />
  <meta name=\\"description\\" content=\\"A simple static website generator\\" />
</head>

<body>
  <header>
  <div class=\\"header-logo-panel\\">
    <div class=\\"header-logo\\">
      <div class=\\"header-image\\">
        <a href=\\"/prefix/\\">
          <img src=\\"/prefix/images/logo-128.png\\" alt=\\"Koala Logo\\" />
        </a>
      </div>
      <div class=\\"header-text\\">
        <div class=\\"header-title\\">
          <a href=\\"/prefix/\\">Koala</a>
        </div>
        <div class=\\"header-subtitle\\">
          <a href=\\"/prefix/\\">Static Website Generator</a>
        </div>
      </div>
    </div>
  </div>

  <nav>
    <ul>
      <li><a href=\\"/prefix/about.html\\">About</a></li>
    </ul>
  </nav>
</header>
  <main>
  <div class=\\"pages\\">
    <h1>Tag hello</h1>
    <section>
      <h2>
        <a href=\\"/prefix/hello-world.html\\">Hello World</a>
      </h2>
      <div class=\\"post-date\\">20 June 2020</div>
      <p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris maximus orci ac tortor consequat, nec ultrices metus
  dictum. Cras non rutrum dui, ac euismod lacus. Fusce sit amet purus justo. Ut rutrum, metus porttitor pretium aliquet,
  diam dolor fermentum nibh, tempus pretium mi justo in velit. Vestibulum porttitor erat ut congue accumsan. Aenean eu
  lacus quam. Morbi consequat ante quis eros iaculis tempus. Orci varius natoque penatibus et magnis dis parturient
  montes, nascetur ridiculus mus. Nulla sit amet urna vitae elit iaculis blandit. Mauris dignissim consectetur augue vel
  varius.
</p>

      <a href=\\"/prefix/hello-world.html\\">Read More</a>
    </section>
  </div>
  <nav class=\\"pages-navigation\\">
    <div class=\\"page-previous\\">
    </div>
    <div class=\\"page-number\\">Page 1</div>
    <div class=\\"page-next\\">
    </div>
  </nav>



</main>
  <div class=\\"navigation\\">
</div>

<aside>
  <h2>Archives</h2>
  <ul>
    <li>
      <a href=\\"/prefix/archive/2020-06/page-1.html\\">June 2020</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-03/page-1.html\\">March 2019</a>
    </li>
    <li>
      <a href=\\"/prefix/archive/2019-01/page-1.html\\">January 2019</a>
    </li>
  </ul>
</aside>

<aside>
  <h2>Tags</h2>
  <ul>
    <li>
      <a href=\\"/prefix/tags/hello/page-1.html\\">hello</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/intro/page-1.html\\">intro</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/number/page-1.html\\">number</a>
    </li>
    <li>
      <a href=\\"/prefix/tags/test/page-1.html\\">test</a>
    </li>
  </ul>
</aside>

  <footer>
  <b>Koala</b><br />
  Static Website Generator<br />
  <a href=\\"https://www.example.com/prefix/\\">https://www.example.com/prefix/</a>
</footer>
</body>

</html>"
`;
